File: packages/server-laravel/src/PasskeyController.php
<?php

declare(strict_types=1);

namespace OpenPasskey\Laravel;

use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Routing\Controller;
use OpenPasskey\Server\PasskeyError;
use OpenPasskey\Server\PasskeyHandler;

class PasskeyController extends Controller
{
    public function __construct(
        private readonly PasskeyHandler $handler,
    ) {}

    public function beginRegistration(Request $request): JsonResponse
    {
        $body = $request->json()->all();
        try {
            return response()->json($this->handler->beginRegistration(
                $body['userId'] ?? '',
                $body['username'] ?? '',
            ));
        } catch (PasskeyError $e) {
            return response()->json(['error' => $e->getMessage()], $e->statusCode);
        }
    }

    public function finishRegistration(Request $request): JsonResponse
    {
        $body = $request->json()->all();
        try {
            $result = $this->handler->finishRegistration(
                $body['userId'] ?? '',
                $body['credential'] ?? [],
                $body['prfSupported'] ?? false,
            );
        } catch (PasskeyError $e) {
            return response()->json(['error' => $e->getMessage()], $e->statusCode);
        }

        return response()->json($result);
    }

    public function beginAuthentication(Request $request): JsonResponse
    {
        $body = $request->json()->all();
        try {
            return response()->json($this->handler->beginAuthentication($body['userId'] ?? ''));
        } catch (PasskeyError $e) {
            return response()->json(['error' => $e->getMessage()], $e->statusCode);
        }
    }

    public function finishAuthentication(Request $request): JsonResponse
    {
        $body = $request->json()->all();
        try {
            $result = $this->handler->finishAuthentication(
                $body['userId'] ?? '',
                $body['credential'] ?? [],
            );
        } catch (PasskeyError $e) {
            return response()->json(['error' => $e->getMessage()], $e->statusCode);
        }

        return response()->json($result);
    }
}

File: packages/server-rails/app/controllers/open_passkey/passkey_controller.rb
# frozen_string_literal: true

module OpenPasskey
  class PasskeyController < ActionController::API
    def begin_registration
      body = JSON.parse(request.body.read)
      result = handler.begin_registration(body["userId"], body["username"])
      render json: result
    rescue PasskeyError => e
      render json: { error: e.message }, status: e.status_code
    end

    def finish_registration
      body = JSON.parse(request.body.read)
      prf = body.dig("credential", "clientExtensionResults", "prf", "enabled") || false
      result = handler.finish_registration(body["userId"], body["credential"], prf)
      render json: result
    rescue PasskeyError => e
      render json: { error: e.message }, status: e.status_code
    end

    def begin_authentication
      body = JSON.parse(request.body.read)
      result = handler.begin_authentication(body["userId"] || "")
      render json: result
    rescue PasskeyError => e
      render json: { error: e.message }, status: e.status_code
    end

    def finish_authentication
      body = JSON.parse(request.body.read)
      result = handler.finish_authentication(body["userId"], body["credential"])
      render json: result
    rescue PasskeyError => e
      render json: { error: e.message }, status: e.status_code
    end

    private

    def handler
      @handler ||= OpenPasskey.handler
    end

  end
end
